test/macros_test.exs

```diff
@@ -97,38 +97,61 @@ defmodule Bureaucrat.MacrosTest do
     end
   end
 
-  # We want to raise an error when Phoenix.ConnTest macros are called outside a `test` block.
-  # This test mimics a userland test that calls Phoenix.ConnTest.{get, post, put, delete}
-  # and asserts that a RuntimeError is raised when compiling the code.
-  # This is counterintuitive, but remember that some of Bureaucrat's runtime is during test compilation.
-  for method <- [:get, :post, :put, :delete] do
-    @method method
-    test "Raises runtime error with message when calling #{@method} outside a `test` block" do
-      module_name = :"RuntimeErrorTest#{@method |> Atom.to_string() |> Macro.camelize()}"
-
-      ast =
-        quote do
-          defmodule unquote(module_name) do
-            use ExUnit.Case, async: false
-            import Phoenix.ConnTest, only: :functions
-            import Bureaucrat.Helpers
-            import Bureaucrat.Macros
+  describe "description" do
+    test "is generated from a test description", context do
+      get(context.conn, "/hello")
+      [conn] = Recorder.get_records()
+      assert conn.assigns.bureaucrat_desc == "description is generated from a test description"
+    end
 
-            @endpoint false
+    test "is generated when the request is made from another function", context do
+      hello_request(context.conn)
+      [conn] = Recorder.get_records()
+      assert conn.assigns.bureaucrat_desc == "description is generated when the request is made from another function"
+    end
 
-            test "won't compile because it calls Phoenix.ConnTest.#{unquote(@method)} in private function" do
-              private_caller()
-            end
+    test "is generated when the request is made from another task", context do
+      description =
+        Task.async(fn ->
+          hello_request(context.conn)
+          [conn] = Recorder.get_records()
+          conn.assigns.bureaucrat_desc
+        end)
+        |> Task.await()
 
-            defp private_caller() do
-              unquote(@method)(:fake_conn, "/hello", %{foo: "bar"})
-            end
-          end
-        end
+      assert description == "description is generated when the request is made from another task"
+    end
 
-      error = assert_raise RuntimeError, fn -> Code.eval_quoted(ast) end
+    test "is not auto-generated when one is provided", context do
+      hello_request(context.conn, description: "custom description")
+      [conn] = Recorder.get_records()
+      assert conn.assigns.bureaucrat_desc == "custom description"
+    end
 
-      assert error.message =~ "It looks like you called a `Phoenix.ConnTest` macro inside `private_caller`."
+    defp hello_request(conn, opts \\ []) do
+      case Keyword.fetch(opts, :description) do
+        :error -> get(conn, "/hello")
+        {:ok, description} -> conn |> get_undocumented("/hello") |> doc(description: description)
+      end
```

Comment:
🤦 

So you're telling me that thing I did with the `defp` inside a WHOLE MODULE inside a `Code.eval_quoted` can just be replaced with a `defp` here? That makes sense. My solution looks so convoluted compared to this.